src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -181,234 +209,248 @@ private void CreateStreamAndStartWatcher()
-                            NotifyRenameEventArgs(WatcherChangeTypes.Renamed, newPathRelativeName, relativePath);
 
-                            // Create a new list, if necessary, and add the event
-                            if (handledRenameEvents == null)
-                            {
-                                handledRenameEvents = new List<FSEventStreamEventId>();
-                            }
-                            handledRenameEvents.Add(eventIds[pairedId]);
-                        }
-                    }
-                    else
-                    {
-                        // OS X is wonky where it can give back kFSEventStreamEventFlagItemCreated and kFSEventStreamEventFlagItemRemoved
-                        // for the same item. The only option we have is to stat and see if the item exists; if so send created, otherwise send deleted.
-                        if ((IsFlagSet(eventFlags[i], Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemCreated)) ||
-                            (IsFlagSet(eventFlags[i], Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRemoved)))
+                        // Check if this is a rename
+                        if (IsFlagSet(eventFlags[i], Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRenamed))
                         {
-                            if (DoesItemExist(eventPaths[i], IsFlagSet(eventFlags[i], Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsFile)))
+                            // Find the rename that is paired to this rename, which should be the next rename in the list
+                            long pairedId = FindRenameChangePairedChange(i, eventPaths, eventFlags, eventIds);
+                            if (pairedId == long.MinValue)
                             {
-                                NotifyFileSystemEventArgs(WatcherChangeTypes.Created, relativePath);
+                                // Getting here means we have a rename without a pair, meaning it should be a create for the 
+                                // move from unwatched folder to watcher folder scenario or a move from the watcher folder out.
+                                // Check if the item exists on disk to check which it is
+                                if (DoesItemExist(eventPaths[i], IsFlagSet(eventFlags[i], Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsFile)))
+                                {
+                                    watcher.NotifyFileSystemEventArgs(WatcherChangeTypes.Created, relativePath);
+                                }
+                                else
+                                {
+                                    watcher.NotifyFileSystemEventArgs(WatcherChangeTypes.Deleted, relativePath);
+                                }
```

Comment:
Nit: similarly, this could be:
```C#
var wct = DoesItemExist(eventPaths[i], IsFlagSet(eventFlags[i], Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsFile) ? WatcherChangeTypes.Created : WatcherChangeTypes.Deleted;
watcher.NotifyFilSystemEventArgs(wct, relativePath);
```